src/haz3lcore/statics/Statics.re

```diff
@@ -47,13 +47,26 @@ let map_m = (f, xs, m: Map.t) =>
 let add_info = (ids: list(Id.t), info: Info.t, m: Map.t): Map.t =>
   ids |> List.fold_left((m, id) => Id.Map.add(id, info, m), m);
 
-let extend_let_def_ctx =
-    (ctx: Ctx.t, pat: UPat.t, pat_ctx: Ctx.t, def: UExp.t): Ctx.t =>
-  if (UPat.is_tuple_of_arrows(pat) && UExp.is_tuple_of_functions(def)) {
-    pat_ctx;
-  } else {
-    ctx;
+let is_recursive = (ctx, p, def, syn: Typ.t) => {
+  switch (Term.UPat.get_num_of_vars(p), Term.UExp.get_num_of_functions(def)) {
+  | (Some(num_vars), Some(num_fns))
+      when num_vars != 0 && num_vars == num_fns =>
+    switch (Typ.normalize(ctx, syn)) {
+    | Unknown(_)
+    | Arrow(_) => num_vars == 1
```

Comment:
Annotations for recursive functions (normalized) can be: `Arrow`s, `Invalid`s, `MultiHole`s, `TypeHole`s, and free type variables.

Reply:
I think we want this to work for any type that has a matched arrow type, so that should include Invalid/MultiHole/free type variables. Perhaps you can use matched arrow type directly after normalizing?

Reply:
I think we have already been supporting this:
![image](https://github.com/hazelgrove/hazel/assets/108375845/1874ae13-66e2-4a09-b7be-38c6e65b4f01)